src/passes/SimplifyGlobals.cpp

```diff
@@ -50,10 +50,25 @@ namespace wasm {
 namespace {
 
 struct GlobalInfo {
+  // Whether the global is imported and exported.
   bool imported = false;
   bool exported = false;
-  std::atomic<bool> written;
-  std::atomic<bool> read;
+
+  // How many times the global is written and read.
+  std::atomic<Index> written{0};
+  std::atomic<Index> read{0};
+
+  // How many times the global is "read, but only to write", that is, is used in
+  // this pattern:
+  //
+  //   if (global == X) { global = Y }
+  //
+  // where X and Y have no side effects. If all we have are such reads only to
```

Comment:
Why is it required that X and Y have no side effects? We are not reordering anything, so as long as the concerned global is read only to write, can't other patterns with other orthogonal side effects be optimized too?

Or, is this pattern without side effects just an easy and common enough to detect and optimize away? In that case, maybe better to elaborate that side effects don't necessary prevent the optimization but they are not the common pattern or something...

Reply:
```js
if (global == 0) {
  global = 1;
  print("first time");
}
```
That side effect will be noticed, as it's printed the first time we reach it. So we can't remove writes to the global, as then we'd end up with
```js
// The read turns into 0, the default value, which is never
// written so it is constant.
if (0 == 0) {
  // The write that was here was removed.
  print("first time");
}
```
That will never print.